[BUG] mismatch cpu and gpu result in test_aqe_join_reused_exchange_inequality_condition failed

**Describe the bug**
We saw this failures in test run w/ multiple-nodes UCX ENV, cloudera ENV and yarn ENV

the case was introduced in https://github.com/NVIDIA/spark-rapids/pull/7110

```
[2022-11-24T15:10:38.487Z] FAILED ../../src/main/python/aqe_test.py::test_aqe_join_reused_exchange_inequality_condition[inner][IGNORE_ORDER({'local': True}), ALLOW_NON_GPU(BroadcastNestedLoopJoinExec,BroadcastExchangeExec,Cast,DateSub)]

[2022-11-24T15:10:38.487Z] FAILED ../../src/main/python/aqe_test.py::test_aqe_join_reused_exchange_inequality_condition[cross][IGNORE_ORDER({'local': True}), ALLOW_NON_GPU(BroadcastNestedLoopJoinExec,BroadcastExchangeExec,Cast,DateSub)]
```

```
[2022-11-24T15:10:38.486Z] =================================== FAILURES ===================================

[2022-11-24T15:10:38.486Z] __________ test_aqe_join_reused_exchange_inequality_condition[inner] ___________

[2022-11-24T15:10:38.486Z] 

[2022-11-24T15:10:38.486Z] spark_tmp_path = '/tmp/pyspark_tests//rl-r7525-d32-u27.raplab.nvidia.com-main-4748-2106370104/'

[2022-11-24T15:10:38.486Z] join = 'inner'

[2022-11-24T15:10:38.486Z] 

[2022-11-24T15:10:38.486Z]     @ignore_order(local=True)

[2022-11-24T15:10:38.486Z]     @allow_non_gpu('BroadcastNestedLoopJoinExec', 'BroadcastExchangeExec', 'Cast', 'DateSub')

[2022-11-24T15:10:38.486Z]     @pytest.mark.parametrize('join', joins, ids=idfn)

[2022-11-24T15:10:38.486Z]     def test_aqe_join_reused_exchange_inequality_condition(spark_tmp_path, join):

[2022-11-24T15:10:38.486Z]         data_path = spark_tmp_path + '/PARQUET_DATA'

[2022-11-24T15:10:38.486Z]         def prep(spark):

[2022-11-24T15:10:38.486Z]             data = [

[2022-11-24T15:10:38.486Z]                 (("Adam ", "", "Green"), "1", "M", 1000),

[2022-11-24T15:10:38.486Z]                 (("Bob ", "Middle", "Green"), "2", "M", 2000),

[2022-11-24T15:10:38.486Z]                 (("Cathy ", "", "Green"), "3", "F", 3000)

[2022-11-24T15:10:38.486Z]             ]

[2022-11-24T15:10:38.486Z]             schema = (StructType()

[2022-11-24T15:10:38.486Z]                       .add("name", StructType()

[2022-11-24T15:10:38.486Z]                            .add("firstname", StringType())

[2022-11-24T15:10:38.486Z]                            .add("middlename", StringType())

[2022-11-24T15:10:38.486Z]                            .add("lastname", StringType()))

[2022-11-24T15:10:38.486Z]                       .add("id", StringType())

[2022-11-24T15:10:38.486Z]                       .add("gender", StringType())

[2022-11-24T15:10:38.486Z]                       .add("salary", IntegerType()))

[2022-11-24T15:10:38.486Z]     

[2022-11-24T15:10:38.486Z]             df = spark.createDataFrame(spark.sparkContext.parallelize(data),schema)

[2022-11-24T15:10:38.486Z]             df2 = df.withColumn("dt",current_date().alias("dt")).withColumn("ts",current_timestamp().alias("ts"))

[2022-11-24T15:10:38.486Z]     

[2022-11-24T15:10:38.486Z]             df2.printSchema

[2022-11-24T15:10:38.486Z]             df2.write.format("parquet").mode("overwrite").save(data_path)

[2022-11-24T15:10:38.486Z]     

[2022-11-24T15:10:38.486Z]         with_cpu_session(prep)

[2022-11-24T15:10:38.486Z]     

[2022-11-24T15:10:38.486Z]         def do_it(spark):

[2022-11-24T15:10:38.486Z]             newdf2 = spark.read.parquet(data_path)

[2022-11-24T15:10:38.486Z]             newdf2.createOrReplaceTempView("df2")

[2022-11-24T15:10:38.486Z]     

[2022-11-24T15:10:38.486Z]             return spark.sql(

[2022-11-24T15:10:38.486Z]                 """

[2022-11-24T15:10:38.486Z]                 select *

[2022-11-24T15:10:38.486Z]                     from (

[2022-11-24T15:10:38.486Z]                         select distinct a.salary

[2022-11-24T15:10:38.486Z]                         from df2 a {join} join (select max(date(ts)) as state_start from df2) b

[2022-11-24T15:10:38.486Z]                         on date(a.ts) > b.state_start - 2)

[2022-11-24T15:10:38.486Z]                     where salary in (

[2022-11-24T15:10:38.486Z]                         select salary from (select a.salary

[2022-11-24T15:10:38.486Z]                         from df2 a inner join (select max(date(ts)) as state_start from df2) b on date(a.ts) > b.state_start - 2

[2022-11-24T15:10:38.486Z]                         limit 1))

[2022-11-24T15:10:38.486Z]                 """.format(join=join)

[2022-11-24T15:10:38.486Z]             )

[2022-11-24T15:10:38.486Z]     

[2022-11-24T15:10:38.486Z] >       assert_gpu_and_cpu_are_equal_collect(do_it, conf=_adaptive_conf)

[2022-11-24T15:10:38.486Z] 

[2022-11-24T15:10:38.486Z] ../../src/main/python/aqe_test.py:185: 

[2022-11-24T15:10:38.486Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

[2022-11-24T15:10:38.486Z] ../../src/main/python/asserts.py:548: in assert_gpu_and_cpu_are_equal_collect

[2022-11-24T15:10:38.486Z]     _assert_gpu_and_cpu_are_equal(func, 'COLLECT', conf=conf, is_cpu_first=is_cpu_first)

[2022-11-24T15:10:38.486Z] ../../src/main/python/asserts.py:479: in _assert_gpu_and_cpu_are_equal

[2022-11-24T15:10:38.486Z]     assert_equal(from_cpu, from_gpu)

[2022-11-24T15:10:38.486Z] ../../src/main/python/asserts.py:106: in assert_equal

[2022-11-24T15:10:38.486Z]     _assert_equal(cpu, gpu, float_check=get_float_check(), path=[])

[2022-11-24T15:10:38.486Z] ../../src/main/python/asserts.py:42: in _assert_equal

[2022-11-24T15:10:38.486Z]     _assert_equal(cpu[index], gpu[index], float_check, path + [index])

[2022-11-24T15:10:38.486Z] ../../src/main/python/asserts.py:35: in _assert_equal

[2022-11-24T15:10:38.486Z]     _assert_equal(cpu[field], gpu[field], float_check, path + [field])

[2022-11-24T15:10:38.486Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

[2022-11-24T15:10:38.486Z] 

[2022-11-24T15:10:38.486Z] cpu = 1000, gpu = 2000

[2022-11-24T15:10:38.486Z] float_check = <function get_float_check.<locals>.<lambda> at 0x7f1e706ac430>

[2022-11-24T15:10:38.486Z] path = [0, 'salary']

[2022-11-24T15:10:38.486Z] 

[2022-11-24T15:10:38.486Z]     def _assert_equal(cpu, gpu, float_check, path):

[2022-11-24T15:10:38.486Z]         t = type(cpu)

[2022-11-24T15:10:38.486Z]         if (t is Row):

[2022-11-24T15:10:38.486Z]             assert len(cpu) == len(gpu), "CPU and GPU row have different lengths at {} CPU: {} GPU: {}".format(path, len(cpu), len(gpu))

[2022-11-24T15:10:38.486Z]             if hasattr(cpu, "__fields__") and hasattr(gpu, "__fields__"):

[2022-11-24T15:10:38.486Z]                 assert cpu.__fields__ == gpu.__fields__, "CPU and GPU row have different fields at {} CPU: {} GPU: {}".format(path, cpu.__fields__, gpu.__fields__)

[2022-11-24T15:10:38.486Z]                 for field in cpu.__fields__:

[2022-11-24T15:10:38.486Z]                     _assert_equal(cpu[field], gpu[field], float_check, path + [field])

[2022-11-24T15:10:38.486Z]             else:

[2022-11-24T15:10:38.486Z]                 for index in range(len(cpu)):

[2022-11-24T15:10:38.486Z]                     _assert_equal(cpu[index], gpu[index], float_check, path + [index])

[2022-11-24T15:10:38.486Z]         elif (t is list):

[2022-11-24T15:10:38.486Z]             assert len(cpu) == len(gpu), "CPU and GPU list have different lengths at {} CPU: {} GPU: {}".format(path, len(cpu), len(gpu))

[2022-11-24T15:10:38.486Z]             for index in range(len(cpu)):

[2022-11-24T15:10:38.486Z]                 _assert_equal(cpu[index], gpu[index], float_check, path + [index])

[2022-11-24T15:10:38.486Z]         elif (t is tuple):

[2022-11-24T15:10:38.486Z]             assert len(cpu) == len(gpu), "CPU and GPU list have different lengths at {} CPU: {} GPU: {}".format(path, len(cpu), len(gpu))

[2022-11-24T15:10:38.486Z]             for index in range(len(cpu)):

[2022-11-24T15:10:38.486Z]                 _assert_equal(cpu[index], gpu[index], float_check, path + [index])

[2022-11-24T15:10:38.486Z]         elif (t is pytypes.GeneratorType):

[2022-11-24T15:10:38.486Z]             index = 0

[2022-11-24T15:10:38.486Z]             # generator has no zip :( so we have to do this the hard way

[2022-11-24T15:10:38.486Z]             done = False

[2022-11-24T15:10:38.486Z]             while not done:

[2022-11-24T15:10:38.486Z]                 sub_cpu = None

[2022-11-24T15:10:38.486Z]                 sub_gpu = None

[2022-11-24T15:10:38.486Z]                 try:

[2022-11-24T15:10:38.486Z]                     sub_cpu = next(cpu)

[2022-11-24T15:10:38.486Z]                 except StopIteration:

[2022-11-24T15:10:38.486Z]                     done = True

[2022-11-24T15:10:38.486Z]     

[2022-11-24T15:10:38.486Z]                 try:

[2022-11-24T15:10:38.486Z]                     sub_gpu = next(gpu)

[2022-11-24T15:10:38.486Z]                 except StopIteration:

[2022-11-24T15:10:38.486Z]                     done = True

[2022-11-24T15:10:38.486Z]     

[2022-11-24T15:10:38.486Z]                 if done:

[2022-11-24T15:10:38.486Z]                     assert sub_cpu == sub_gpu and sub_cpu == None, "CPU and GPU generators have different lengths at {}".format(path)

[2022-11-24T15:10:38.486Z]                 else:

[2022-11-24T15:10:38.486Z]                     _assert_equal(sub_cpu, sub_gpu, float_check, path + [index])

[2022-11-24T15:10:38.486Z]     

[2022-11-24T15:10:38.486Z]                 index = index + 1

[2022-11-24T15:10:38.486Z]         elif (t is dict):

[2022-11-24T15:10:38.486Z]             # The order of key/values is not guaranteed in python dicts, nor are they guaranteed by Spark

[2022-11-24T15:10:38.486Z]             # so sort the items to do our best with ignoring the order of dicts

[2022-11-24T15:10:38.486Z]             cpu_items = list(cpu.items()).sort(key=_RowCmp)

[2022-11-24T15:10:38.486Z]             gpu_items = list(gpu.items()).sort(key=_RowCmp)

[2022-11-24T15:10:38.486Z]             _assert_equal(cpu_items, gpu_items, float_check, path + ["map"])

[2022-11-24T15:10:38.486Z]         elif (t is int):

[2022-11-24T15:10:38.486Z] >           assert cpu == gpu, "GPU and CPU int values are different at {}".format(path)

[2022-11-24T15:10:38.486Z] E           AssertionError: GPU and CPU int values are different at [0, 'salary']

[2022-11-24T15:10:38.486Z] 

[2022-11-24T15:10:38.486Z] ../../src/main/python/asserts.py:77: AssertionError

[2022-11-24T15:10:38.486Z] ----------------------------- Captured stdout call -----------------------------

[2022-11-24T15:10:38.486Z] ### CPU RUN ###

[2022-11-24T15:10:38.486Z] ### GPU RUN ###

[2022-11-24T15:10:38.486Z] ### COLLECT: GPU TOOK 0.804142951965332 CPU TOOK 0.8476741313934326 ###

[2022-11-24T15:10:38.486Z] CPU OUTPUT: [Row(salary=1000)]

[2022-11-24T15:10:38.486Z] GPU OUTPUT: [Row(salary=2000)]

[2022-11-24T15:10:38.486Z] __________ test_aqe_join_reused_exchange_inequality_condition[cross] ___________

[2022-11-24T15:10:38.486Z] 

[2022-11-24T15:10:38.486Z] spark_tmp_path = '/tmp/pyspark_tests//rl-r7525-d32-u27.raplab.nvidia.com-main-4748-656661051/'

[2022-11-24T15:10:38.486Z] join = 'cross'

[2022-11-24T15:10:38.486Z] 

[2022-11-24T15:10:38.486Z]     @ignore_order(local=True)

[2022-11-24T15:10:38.486Z]     @allow_non_gpu('BroadcastNestedLoopJoinExec', 'BroadcastExchangeExec', 'Cast', 'DateSub')

[2022-11-24T15:10:38.486Z]     @pytest.mark.parametrize('join', joins, ids=idfn)

[2022-11-24T15:10:38.486Z]     def test_aqe_join_reused_exchange_inequality_condition(spark_tmp_path, join):

[2022-11-24T15:10:38.486Z]         data_path = spark_tmp_path + '/PARQUET_DATA'

[2022-11-24T15:10:38.486Z]         def prep(spark):

[2022-11-24T15:10:38.486Z]             data = [

[2022-11-24T15:10:38.486Z]                 (("Adam ", "", "Green"), "1", "M", 1000),

[2022-11-24T15:10:38.486Z]                 (("Bob ", "Middle", "Green"), "2", "M", 2000),

[2022-11-24T15:10:38.486Z]                 (("Cathy ", "", "Green"), "3", "F", 3000)

[2022-11-24T15:10:38.486Z]             ]

[2022-11-24T15:10:38.486Z]             schema = (StructType()

[2022-11-24T15:10:38.486Z]                       .add("name", StructType()

[2022-11-24T15:10:38.486Z]                            .add("firstname", StringType())

[2022-11-24T15:10:38.486Z]                            .add("middlename", StringType())

[2022-11-24T15:10:38.486Z]                            .add("lastname", StringType()))

[2022-11-24T15:10:38.486Z]                       .add("id", StringType())

[2022-11-24T15:10:38.486Z]                       .add("gender", StringType())

[2022-11-24T15:10:38.486Z]                       .add("salary", IntegerType()))

[2022-11-24T15:10:38.486Z]     

[2022-11-24T15:10:38.486Z]             df = spark.createDataFrame(spark.sparkContext.parallelize(data),schema)

[2022-11-24T15:10:38.486Z]             df2 = df.withColumn("dt",current_date().alias("dt")).withColumn("ts",current_timestamp().alias("ts"))

[2022-11-24T15:10:38.486Z]     

[2022-11-24T15:10:38.486Z]             df2.printSchema

[2022-11-24T15:10:38.486Z]             df2.write.format("parquet").mode("overwrite").save(data_path)

[2022-11-24T15:10:38.486Z]     

[2022-11-24T15:10:38.486Z]         with_cpu_session(prep)

[2022-11-24T15:10:38.486Z]     

[2022-11-24T15:10:38.486Z]         def do_it(spark):

[2022-11-24T15:10:38.486Z]             newdf2 = spark.read.parquet(data_path)

[2022-11-24T15:10:38.486Z]             newdf2.createOrReplaceTempView("df2")

[2022-11-24T15:10:38.486Z]     

[2022-11-24T15:10:38.486Z]             return spark.sql(

[2022-11-24T15:10:38.486Z]                 """

[2022-11-24T15:10:38.486Z]                 select *

[2022-11-24T15:10:38.486Z]                     from (

[2022-11-24T15:10:38.486Z]                         select distinct a.salary

[2022-11-24T15:10:38.486Z]                         from df2 a {join} join (select max(date(ts)) as state_start from df2) b

[2022-11-24T15:10:38.486Z]                         on date(a.ts) > b.state_start - 2)

[2022-11-24T15:10:38.486Z]                     where salary in (

[2022-11-24T15:10:38.486Z]                         select salary from (select a.salary

[2022-11-24T15:10:38.486Z]                         from df2 a inner join (select max(date(ts)) as state_start from df2) b on date(a.ts) > b.state_start - 2

[2022-11-24T15:10:38.486Z]                         limit 1))

[2022-11-24T15:10:38.486Z]                 """.format(join=join)

[2022-11-24T15:10:38.486Z]             )

[2022-11-24T15:10:38.486Z]     

[2022-11-24T15:10:38.486Z] >       assert_gpu_and_cpu_are_equal_collect(do_it, conf=_adaptive_conf)

[2022-11-24T15:10:38.486Z] 

[2022-11-24T15:10:38.486Z] ../../src/main/python/aqe_test.py:185: 

[2022-11-24T15:10:38.486Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

[2022-11-24T15:10:38.486Z] ../../src/main/python/asserts.py:548: in assert_gpu_and_cpu_are_equal_collect

[2022-11-24T15:10:38.486Z]     _assert_gpu_and_cpu_are_equal(func, 'COLLECT', conf=conf, is_cpu_first=is_cpu_first)

[2022-11-24T15:10:38.486Z] ../../src/main/python/asserts.py:479: in _assert_gpu_and_cpu_are_equal

[2022-11-24T15:10:38.486Z]     assert_equal(from_cpu, from_gpu)

[2022-11-24T15:10:38.486Z] ../../src/main/python/asserts.py:106: in assert_equal

[2022-11-24T15:10:38.486Z]     _assert_equal(cpu, gpu, float_check=get_float_check(), path=[])

[2022-11-24T15:10:38.486Z] ../../src/main/python/asserts.py:42: in _assert_equal

[2022-11-24T15:10:38.486Z]     _assert_equal(cpu[index], gpu[index], float_check, path + [index])

[2022-11-24T15:10:38.486Z] ../../src/main/python/asserts.py:35: in _assert_equal

[2022-11-24T15:10:38.486Z]     _assert_equal(cpu[field], gpu[field], float_check, path + [field])

[2022-11-24T15:10:38.486Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

[2022-11-24T15:10:38.486Z] 

[2022-11-24T15:10:38.486Z] cpu = 2000, gpu = 1000

[2022-11-24T15:10:38.486Z] float_check = <function get_float_check.<locals>.<lambda> at 0x7f1e5eae8a60>

[2022-11-24T15:10:38.486Z] path = [0, 'salary']

[2022-11-24T15:10:38.486Z] 

[2022-11-24T15:10:38.486Z]     def _assert_equal(cpu, gpu, float_check, path):

[2022-11-24T15:10:38.486Z]         t = type(cpu)

[2022-11-24T15:10:38.486Z]         if (t is Row):

[2022-11-24T15:10:38.486Z]             assert len(cpu) == len(gpu), "CPU and GPU row have different lengths at {} CPU: {} GPU: {}".format(path, len(cpu), len(gpu))

[2022-11-24T15:10:38.486Z]             if hasattr(cpu, "__fields__") and hasattr(gpu, "__fields__"):

[2022-11-24T15:10:38.486Z]                 assert cpu.__fields__ == gpu.__fields__, "CPU and GPU row have different fields at {} CPU: {} GPU: {}".format(path, cpu.__fields__, gpu.__fields__)

[2022-11-24T15:10:38.486Z]                 for field in cpu.__fields__:

[2022-11-24T15:10:38.486Z]                     _assert_equal(cpu[field], gpu[field], float_check, path + [field])

[2022-11-24T15:10:38.486Z]             else:

[2022-11-24T15:10:38.486Z]                 for index in range(len(cpu)):

[2022-11-24T15:10:38.486Z]                     _assert_equal(cpu[index], gpu[index], float_check, path + [index])

[2022-11-24T15:10:38.486Z]         elif (t is list):

[2022-11-24T15:10:38.486Z]             assert len(cpu) == len(gpu), "CPU and GPU list have different lengths at {} CPU: {} GPU: {}".format(path, len(cpu), len(gpu))

[2022-11-24T15:10:38.486Z]             for index in range(len(cpu)):

[2022-11-24T15:10:38.486Z]                 _assert_equal(cpu[index], gpu[index], float_check, path + [index])

[2022-11-24T15:10:38.486Z]         elif (t is tuple):

[2022-11-24T15:10:38.486Z]             assert len(cpu) == len(gpu), "CPU and GPU list have different lengths at {} CPU: {} GPU: {}".format(path, len(cpu), len(gpu))

[2022-11-24T15:10:38.486Z]             for index in range(len(cpu)):

[2022-11-24T15:10:38.486Z]                 _assert_equal(cpu[index], gpu[index], float_check, path + [index])

[2022-11-24T15:10:38.486Z]         elif (t is pytypes.GeneratorType):

[2022-11-24T15:10:38.486Z]             index = 0

[2022-11-24T15:10:38.486Z]             # generator has no zip :( so we have to do this the hard way

[2022-11-24T15:10:38.487Z]             done = False

[2022-11-24T15:10:38.487Z]             while not done:

[2022-11-24T15:10:38.487Z]                 sub_cpu = None

[2022-11-24T15:10:38.487Z]                 sub_gpu = None

[2022-11-24T15:10:38.487Z]                 try:

[2022-11-24T15:10:38.487Z]                     sub_cpu = next(cpu)

[2022-11-24T15:10:38.487Z]                 except StopIteration:

[2022-11-24T15:10:38.487Z]                     done = True

[2022-11-24T15:10:38.487Z]     

[2022-11-24T15:10:38.487Z]                 try:

[2022-11-24T15:10:38.487Z]                     sub_gpu = next(gpu)

[2022-11-24T15:10:38.487Z]                 except StopIteration:

[2022-11-24T15:10:38.487Z]                     done = True

[2022-11-24T15:10:38.487Z]     

[2022-11-24T15:10:38.487Z]                 if done:

[2022-11-24T15:10:38.487Z]                     assert sub_cpu == sub_gpu and sub_cpu == None, "CPU and GPU generators have different lengths at {}".format(path)

[2022-11-24T15:10:38.487Z]                 else:

[2022-11-24T15:10:38.487Z]                     _assert_equal(sub_cpu, sub_gpu, float_check, path + [index])

[2022-11-24T15:10:38.487Z]     

[2022-11-24T15:10:38.487Z]                 index = index + 1

[2022-11-24T15:10:38.487Z]         elif (t is dict):

[2022-11-24T15:10:38.487Z]             # The order of key/values is not guaranteed in python dicts, nor are they guaranteed by Spark

[2022-11-24T15:10:38.487Z]             # so sort the items to do our best with ignoring the order of dicts

[2022-11-24T15:10:38.487Z]             cpu_items = list(cpu.items()).sort(key=_RowCmp)

[2022-11-24T15:10:38.487Z]             gpu_items = list(gpu.items()).sort(key=_RowCmp)

[2022-11-24T15:10:38.487Z]             _assert_equal(cpu_items, gpu_items, float_check, path + ["map"])

[2022-11-24T15:10:38.487Z]         elif (t is int):

[2022-11-24T15:10:38.487Z] >           assert cpu == gpu, "GPU and CPU int values are different at {}".format(path)

[2022-11-24T15:10:38.487Z] E           AssertionError: GPU and CPU int values are different at [0, 'salary']
```

**Steps/Code to reproduce bug**
Please provide a list of steps or a code sample to reproduce the issue.
Avoid posting private or sensitive data.

After some triage of this bug, it looks like the failure in this test is not specifically from the plugin updates that were added in #7110, but probably somewhere else in the code. I reverted the Scala plugin changes from #7110 but kept the integration test `test_aqe_join_reused_exchange_inequality_condition`, and ran a couple of tests:

From quick and dirty local distributed run: 
```bash
$ NUM_LOCAL_EXECS=2 TEST="test_aqe_join_reused_exchange_inequality_condition" TEST_PARALLEL=1 ./run_pyspark_from_build.sh
...
________ test_aqe_join_reused_exchange_inequality_condition[left semi] _________

spark_tmp_path = '/tmp/pyspark_tests//7f588d3-lcedt-master-1615675-1120437873/'
join = 'left semi'

    @ignore_order(local=True)
    @allow_non_gpu('BroadcastNestedLoopJoinExec', 'BroadcastExchangeExec', 'Cast', 'DateSub')
    @pytest.mark.parametrize('join', joins, ids=idfn)
    def test_aqe_join_reused_exchange_inequality_condition(spark_tmp_path, join):
        data_path = spark_tmp_path + '/PARQUET_DATA'
        def prep(spark):
            data = [
                (("Adam ", "", "Green"), "1", "M", 1000),
                (("Bob ", "Middle", "Green"), "2", "M", 2000),
                (("Cathy ", "", "Green"), "3", "F", 3000)
            ]
            schema = (StructType()
                      .add("name", StructType()
                           .add("firstname", StringType())
                           .add("middlename", StringType())
                           .add("lastname", StringType()))
                      .add("id", StringType())
                      .add("gender", StringType())
                      .add("salary", IntegerType()))

            df = spark.createDataFrame(spark.sparkContext.parallelize(data),schema)
            df2 = df.withColumn("dt",current_date().alias("dt")).withColumn("ts",current_timestamp().alias("ts"))

            df2.printSchema
            df2.write.format("parquet").mode("overwrite").save(data_path)

        with_cpu_session(prep)

        def do_it(spark):
            newdf2 = spark.read.parquet(data_path)
            newdf2.createOrReplaceTempView("df2")

            return spark.sql(
                """
                select *
                    from (
                        select distinct a.salary
                        from df2 a {join} join (select max(date(ts)) as state_start from df2) b
                        on date(a.ts) > b.state_start - 2)
                    where salary in (
                        select salary from (select a.salary
                        from df2 a inner join (select max(date(ts)) as state_start from df2) b on date(a.ts) > b.state_start - 2
                        limit 1))
                """.format(join=join)
            )

        conf = {
            'spark.sql.adaptive.enabled': 'true',
            'spark.rapids.sql.debug.logTransformations': 'true'
        }

>       assert_gpu_and_cpu_are_equal_collect(do_it, conf=conf)

../../src/main/python/aqe_test.py:190:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../src/main/python/asserts.py:548: in assert_gpu_and_cpu_are_equal_collect
    _assert_gpu_and_cpu_are_equal(func, 'COLLECT', conf=conf, is_cpu_first=is_cpu_first)
../../src/main/python/asserts.py:479: in _assert_gpu_and_cpu_are_equal
    assert_equal(from_cpu, from_gpu)
../../src/main/python/asserts.py:106: in assert_equal
    _assert_equal(cpu, gpu, float_check=get_float_check(), path=[])
../../src/main/python/asserts.py:42: in _assert_equal
    _assert_equal(cpu[index], gpu[index], float_check, path + [index])
../../src/main/python/asserts.py:35: in _assert_equal
    _assert_equal(cpu[field], gpu[field], float_check, path + [field])
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

cpu = 2000, gpu = 1000
float_check = <function get_float_check.<locals>.<lambda> at 0x7f699a19cca0>
path = [0, 'salary']

    def _assert_equal(cpu, gpu, float_check, path):
        t = type(cpu)
        if (t is Row):
            assert len(cpu) == len(gpu), "CPU and GPU row have different lengths at {} CPU: {} GPU: {}".format(path, len(cpu), len(gpu))
            if hasattr(cpu, "__fields__") and hasattr(gpu, "__fields__"):
                assert cpu.__fields__ == gpu.__fields__, "CPU and GPU row have different fields at {} CPU: {} GPU: {}".format(path, cpu.__fields__, gpu.__fields__)
                for field in cpu.__fields__:
                    _assert_equal(cpu[field], gpu[field], float_check, path + [field])
            else:
                for index in range(len(cpu)):
                    _assert_equal(cpu[index], gpu[index], float_check, path + [index])
        elif (t is list):
            assert len(cpu) == len(gpu), "CPU and GPU list have different lengths at {} CPU: {} GPU: {}".format(path, len(cpu), len(gpu))
            for index in range(len(cpu)):
                _assert_equal(cpu[index], gpu[index], float_check, path + [index])
        elif (t is tuple):
            assert len(cpu) == len(gpu), "CPU and GPU list have different lengths at {} CPU: {} GPU: {}".format(path, len(cpu), len(gpu))
            for index in range(len(cpu)):
                _assert_equal(cpu[index], gpu[index], float_check, path + [index])
        elif (t is pytypes.GeneratorType):
            index = 0
            # generator has no zip :( so we have to do this the hard way
            done = False
            while not done:
                sub_cpu = None
                sub_gpu = None
                try:
                    sub_cpu = next(cpu)
                except StopIteration:
                    done = True

                try:
                    sub_gpu = next(gpu)
                except StopIteration:
                    done = True

                if done:
                    assert sub_cpu == sub_gpu and sub_cpu == None, "CPU and GPU generators have different lengths at {}".format(path)
                else:
                    _assert_equal(sub_cpu, sub_gpu, float_check, path + [index])

                index = index + 1
        elif (t is dict):
            # The order of key/values is not guaranteed in python dicts, nor are they guaranteed by Spark
            # so sort the items to do our best with ignoring the order of dicts
            cpu_items = list(cpu.items()).sort(key=_RowCmp)
            gpu_items = list(gpu.items()).sort(key=_RowCmp)
            _assert_equal(cpu_items, gpu_items, float_check, path + ["map"])
        elif (t is int):
>           assert cpu == gpu, "GPU and CPU int values are different at {}".format(path)
E           AssertionError: GPU and CPU int values are different at [0, 'salary']

../../src/main/python/asserts.py:77: AssertionError
----------------------------- Captured stdout call -----------------------------
### CPU RUN ###
### GPU RUN ###
### COLLECT: GPU TOOK 0.5961883068084717 CPU TOOK 0.565824031829834 ###
CPU OUTPUT: [Row(salary=2000)]
GPU OUTPUT: [Row(salary=1000)]
...
```

Also, I ran the reverted JAR in YARN with UCX enabled:
```
=================================== FAILURES ===================================
[31m[1m__________ test_aqe_join_reused_exchange_inequality_condition[inner] ___________[0m

spark_tmp_path = '/tmp/pyspark_tests//spark-egx-08-main-96-1027424566/'
join = 'inner'

    @ignore_order(local=True)
    @allow_non_gpu('BroadcastNestedLoopJoinExec', 'BroadcastExchangeExec', 'Cast', 'DateSub')
    @pytest.mark.parametrize('join', joins, ids=idfn)
    def test_aqe_join_reused_exchange_inequality_condition(spark_tmp_path, join):
        data_path = spark_tmp_path + '/PARQUET_DATA'
        def prep(spark):
            data = [
                (("Adam ", "", "Green"), "1", "M", 1000),
                (("Bob ", "Middle", "Green"), "2", "M", 2000),
                (("Cathy ", "", "Green"), "3", "F", 3000)
            ]
            schema = (StructType()
                      .add("name", StructType()
                           .add("firstname", StringType())
                           .add("middlename", StringType())
                           .add("lastname", StringType()))
                      .add("id", StringType())
                      .add("gender", StringType())
                      .add("salary", IntegerType()))
    
            df = spark.createDataFrame(spark.sparkContext.parallelize(data),schema)
            df2 = df.withColumn("dt",current_date().alias("dt")).withColumn("ts",current_timestamp().alias("ts"))
    
            df2.printSchema
            df2.write.format("parquet").mode("overwrite").save(data_path)
    
        with_cpu_session(prep)
    
        def do_it(spark):
            newdf2 = spark.read.parquet(data_path)
            newdf2.createOrReplaceTempView("df2")
    
            return spark.sql(
                """
                select *
                    from (
                        select distinct a.salary
                        from df2 a {join} join (select max(date(ts)) as state_start from df2) b
                        on date(a.ts) > b.state_start - 2)
                    where salary in (
                        select salary from (select a.salary
                        from df2 a inner join (select max(date(ts)) as state_start from df2) b on date(a.ts) > b.state_start - 2
                        limit 1))
                """.format(join=join)
            )
    
        conf = {
            'spark.sql.adaptive.enabled': 'true',
            'spark.rapids.sql.debug.logTransformations': 'true'
        }
    
>       assert_gpu_and_cpu_are_equal_collect(do_it, conf=conf)

[1m[31mintegration_tests/src/main/python/aqe_test.py[0m:190: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[1m[31mintegration_tests/src/main/python/asserts.py[0m:548: in assert_gpu_and_cpu_are_equal_collect
    _assert_gpu_and_cpu_are_equal(func, 'COLLECT', conf=conf, is_cpu_first=is_cpu_first)
[1m[31mintegration_tests/src/main/python/asserts.py[0m:479: in _assert_gpu_and_cpu_are_equal
    assert_equal(from_cpu, from_gpu)
[1m[31mintegration_tests/src/main/python/asserts.py[0m:106: in assert_equal
    _assert_equal(cpu, gpu, float_check=get_float_check(), path=[])
[1m[31mintegration_tests/src/main/python/asserts.py[0m:42: in _assert_equal
    _assert_equal(cpu[index], gpu[index], float_check, path + [index])
[1m[31mintegration_tests/src/main/python/asserts.py[0m:35: in _assert_equal
    _assert_equal(cpu[field], gpu[field], float_check, path + [field])
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cpu = 2000, gpu = 3000
float_check = <function get_float_check.<locals>.<lambda> at 0x7fdec520f6a8>
path = [0, 'salary']

    def _assert_equal(cpu, gpu, float_check, path):
        t = type(cpu)
        if (t is Row):
            assert len(cpu) == len(gpu), "CPU and GPU row have different lengths at {} CPU: {} GPU: {}".format(path, len(cpu), len(gpu))
            if hasattr(cpu, "__fields__") and hasattr(gpu, "__fields__"):
                assert cpu.__fields__ == gpu.__fields__, "CPU and GPU row have different fields at {} CPU: {} GPU: {}".format(path, cpu.__fields__, gpu.__fields__)
                for field in cpu.__fields__:
                    _assert_equal(cpu[field], gpu[field], float_check, path + [field])
            else:
                for index in range(len(cpu)):
                    _assert_equal(cpu[index], gpu[index], float_check, path + [index])
        elif (t is list):
            assert len(cpu) == len(gpu), "CPU and GPU list have different lengths at {} CPU: {} GPU: {}".format(path, len(cpu), len(gpu))
            for index in range(len(cpu)):
                _assert_equal(cpu[index], gpu[index], float_check, path + [index])
        elif (t is tuple):
            assert len(cpu) == len(gpu), "CPU and GPU list have different lengths at {} CPU: {} GPU: {}".format(path, len(cpu), len(gpu))
            for index in range(len(cpu)):
                _assert_equal(cpu[index], gpu[index], float_check, path + [index])
        elif (t is pytypes.GeneratorType):
            index = 0
            # generator has no zip :( so we have to do this the hard way
            done = False
            while not done:
                sub_cpu = None
                sub_gpu = None
                try:
                    sub_cpu = next(cpu)
                except StopIteration:
                    done = True
    
                try:
                    sub_gpu = next(gpu)
                except StopIteration:
                    done = True
    
                if done:
                    assert sub_cpu == sub_gpu and sub_cpu == None, "CPU and GPU generators have different lengths at {}".format(path)
                else:
                    _assert_equal(sub_cpu, sub_gpu, float_check, path + [index])
    
                index = index + 1
        elif (t is dict):
            # The order of key/values is not guaranteed in python dicts, nor are they guaranteed by Spark
            # so sort the items to do our best with ignoring the order of dicts
            cpu_items = list(cpu.items()).sort(key=_RowCmp)
            gpu_items = list(gpu.items()).sort(key=_RowCmp)
            _assert_equal(cpu_items, gpu_items, float_check, path + ["map"])
        elif (t is int):
>           assert cpu == gpu, "GPU and CPU int values are different at {}".format(path)
[1m[31mE           AssertionError: GPU and CPU int values are different at [0, 'salary'][0m

[1m[31mintegration_tests/src/main/python/asserts.py[0m:77: AssertionError
----------------------------- Captured stdout call -----------------------------
### CPU RUN ###
### GPU RUN ###
### COLLECT: GPU TOOK 2.8011553287506104 CPU TOOK 3.404507875442505 ###
CPU OUTPUT: [Row(salary=2000)]
GPU OUTPUT: [Row(salary=3000)]
```

So it looks like this integration test that was added surfaced another bug that affects distributed environments.

--- a/integration_tests/src/main/python/aqe_test.py
+++ b/integration_tests/src/main/python/aqe_test.py
@@ -178,7 +178,7 @@ def test_aqe_join_reused_exchange_inequality_condition(spark_tmp_path, join):
                 where salary in (
                     select salary from (select a.salary
                     from df2 a inner join (select max(date(ts)) as state_start from df2) b on date(a.ts) > b.state_start - 2
-                    limit 1))
+                    order by a.salary limit 1))
             """.format(join=join)
         )
 

